mitmproxy/tools/web/app.py

```diff
@@ -324,6 +327,72 @@ def broadcast(cls, **kwargs):
 class ClientConnection(WebSocketEventBroadcaster):
     connections: ClassVar[set] = set()
 
+    def __init__(self, application: Application, request, **kwargs):
+        super().__init__(application, request, **kwargs)
+        self.application: Application = application
+        self.filters: dict[str, flowfilter.TFilter] = {}  # Instance-level filters
+
+    def send_flow(self, cmd: str, f: mitmproxy.flow.Flow):
+        matches = [
+            {expr.pattern: bool(flowfilter.match(expr, f))}
+            for expr in self.filters.values()
+        ]
+        message = json.dumps(
+            {
+                "resource": "flows",
+                "cmd": cmd,
+                "matches": matches,
+                "data": flow_to_json(f),
+            },
+        ).encode("utf8", "surrogateescape")
+
+        self.send(conn=self, message=message)
+
+    def get_matching_flow_ids(self, name: str) -> List[str]:
+        expr = self.filters.get(name)
+        if not expr:
+            return []
+
+        return [f.id for f in self.application.master.view if expr(f)]
+
+    def send_matching_flow_ids(self, name: str, expr: str):
+        matching_flow_ids = self.get_matching_flow_ids(name)
+
+        message = json.dumps(
+            {
+                "resource": "flows",
+                "cmd": "filtersUpdated",
+                "name": name,
+                "expr": expr,
+                "data": matching_flow_ids,
+            },
+        ).encode("utf8", "surrogateescape")
+
+        self.send(
+            conn=self, message=message
+        )  # send message just to the interested ClientConnection
+
+    async def on_message(self, message: str | bytes):
+        try:
+            data = json.loads(message)
+
+            resource: str = data["resource"]
+            command: str = data["cmd"]
+            name: str = data["name"]
+            expr: str = data["expr"]
+
+            if resource == "flows" and command == "updateFilter":
+                if expr:
+                    self.filters[name] = flowfilter.parse(expr)
+                else:
+                    del self.filters[name]
+
+                self.send_matching_flow_ids(name, expr)
```

Comment:
If you move the first four lines into `send_matching_flow_ids` (we can then rename it to `update_filter`), and then also inline `get_matching_flow_ids`, you can reduce the number of conditionals and jumps:

```python3
def update_filter(self, name: str, expr: str) -> None:
    if expr:
        filt = flowfilter.parse(expr)
        self.filters[name] = filt
        matching_flow_ids = [f.id for f in self.application.master.view if filt(f)]
    else:
        del self.filters[name]
        matching_flow_ids = []
    
    message = json.dumps( 
    # ...
```